core/trino-main/src/main/java/io/trino/util/DateTimeUtils.java

```diff
@@ -67,9 +67,69 @@ public static int parseDate(String value)
         // - the required format is 'YYYY-MM-DD'
         // - all components should be unsigned numbers
         // https://github.com/trinodb/trino/issues/10677
+        int days = tryParseDateYyyyMmDd(value);
+        if (days != Integer.MIN_VALUE) {
+            return days;
+        }
         return toIntExact(TimeUnit.MILLISECONDS.toDays(DATE_FORMATTER.parseMillis(value)));
     }
 
+    /**
+     * Check if the date is in format yyyy-MM-dd, if yes, return the epoch days; otherwise
+     * return Integer.MIN_VALUE, indicating the parse failed, and should fall back to generic
+     * parsing. Mixing the format checking and parsing in one pass for better performance.
+     * @param value date string to be parsed
+     * @return the number of days since 1970-01-01
+     */
+    public static int tryParseDateYyyyMmDd(String value)
+    {
+        if (value.length() != 10) {
+            return Integer.MIN_VALUE;
+        }
+        // year
+        int year = value.charAt(0) - '0';
```

Comment:
Why not: `int year = Integer.parseInt(value, 0, 4, 10)` ? Same when parsing `month` and `day`, I suspect the `Integer.parseInt(CharSequence, beginIndex, endIndex, radix)` overload will perform similarly or better to hand-rolling here but I could be wrong.

Reply:
Integer.parseInt is about 30% slower, probably because there are more checks in that function.

Reply:
I am surprised we can do `Integer.parseInt` faster than Java. is their implementation just suboptimal?
in any case, it's feels better to encapsulate this in a utility method and use that for year, month and day. The code would be much more readable.